llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -2008,6 +2008,30 @@ static bool shouldMergeGEPs(GEPOperator &GEP, GEPOperator &Src) {
   return true;
 }
 
+Instruction *InstCombinerImpl::foldAddLike(Value *LHS, Value *RHS, bool NSW,
```

Comment:
I'd also add `foldAddLikeCommutative()` which is invoked with both LHS/RHS orders, as most add folds with non-constant RHS will want to check both orders.

Then your pattern here will reduce down to:
```
if (match(LHS, m_Sub(m_Value(A), m_Value(B))) &&
    match(RHS, m_Sub(m_Value(C), m_Specific(A)))) {
```